File: 46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        n = len(nums)
        
        def dfs(nums, path):
            #print(nums)
            print(path)
            if n == len(path):
                res.append(path)
                # return # backtracking
            for i in range(len(nums)):
                dfs(nums[:i]+nums[i+1:], path+[nums[i]])
        dfs(nums, [])
        return res

File: 46.cpp
class Solution {
public:
    vector<vector<int>> res;
    
    void helper(int i, vector<int>& nums){
            if(nums.size() == i)
                res.push_back(nums);
            
            for (int j = i; j < nums.size(); ++j){
                swap(nums[i], nums[j]);
                helper(i+1, nums);
                swap(nums[j], nums[i]);
            }
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        helper(0, nums);
        return res;
    }
};